File: src/frontend.cpp
#include "voice_detection/frontend.hpp"

#include "voice_detection/beamforming.hpp"
#include "voice_detection/doa.hpp"
#include "voice_detection/features.hpp"
#include "voice_detection/math_utils.hpp"
#include "voice_detection/signal_enhancement.hpp"

#include <chrono>
#include <cmath>
#include <utility>

namespace voice_detection
{
namespace
{

float estimateClarity(float snr_db)
{
  return clamp(snr_db / 30.0F, 0.0F, 1.0F);
}

float estimateOverlap(const AudioFrame & frame)
{
  if (frame.channels < 2 || frame.frames() == 0) {
    return 0.0F;
  }
  std::vector<float> rms(static_cast<std::size_t>(frame.channels), 0.0F);
  for (int channel = 0; channel < frame.channels; ++channel) {
    double energy = 0.0;
    for (std::size_t i = 0; i < frame.frames(); ++i) {
      const float sample = frame.at(i, channel);
      energy += static_cast<double>(sample) * static_cast<double>(sample);
    }
    rms[static_cast<std::size_t>(channel)] = static_cast<float>(std::sqrt(energy / static_cast<double>(frame.frames())));
  }
  float mean = 0.0F;
  for (const float value : rms) {
    mean += value;
  }
  mean /= static_cast<float>(rms.size());
  float var = 0.0F;
  for (const float value : rms) {
    const float diff = value - mean;
    var += diff * diff;
  }
  var /= static_cast<float>(rms.size());
  const float spread = std::sqrt(var) / std::max(mean, 1e-6F);
  return clamp(1.0F - spread, 0.0F, 1.0F);
}

float estimateSelfEcho(
  const std::vector<float> & mono,
  const std::optional<std::vector<float>> & playback_reference)
{
  if (!playback_reference.has_value()) {
    return 0.0F;
  }
  const std::size_t n = std::min(mono.size(), playback_reference->size());
  if (n < 32) {
    return 0.0F;
  }
  double dot = 0.0;
  double a_energy = 0.0;
  double b_energy = 0.0;
  for (std::size_t i = 0; i < n; ++i) {
    dot += static_cast<double>(mono[i]) * static_cast<double>((*playback_reference)[i]);
    a_energy += static_cast<double>(mono[i]) * static_cast<double>(mono[i]);
    b_energy += static_cast<double>((*playback_reference)[i]) * static_cast<double>((*playback_reference)[i]);
  }
  return clamp(static_cast<float>(std::abs(dot) / (std::sqrt(a_energy * b_energy) + 1e-9)), 0.0F, 1.0F);
}

}  // namespace

CocktailFrontend::CocktailFrontend(MicrophoneProfile profile, int vad_calibration_frames)
: profile_(std::move(profile)),
  pre_vad_(8.0F, -52.0F, vad_calibration_frames),
  post_vad_(8.0F, -52.0F, vad_calibration_frames)
{
}

FrontendOutput CocktailFrontend::process(
  const AudioFrame & frame,
  const std::optional<std::vector<float>> & playback_reference)
{
  AudioFrame phase_safe_frame = frame;
  phase_safe_frame.samples = removeDcPerChannel(frame.samples, frame.channels);
  const auto raw_mono = mixdown(phase_safe_frame);
  const VoiceActivity pre_activity = pre_vad_.process(raw_mono);
  std::optional<float> azimuth;
  float doa_confidence = 0.0F;
  if (pre_activity.active && profile_.supports_doa && !profile_.doa_channel_indices.empty()) {
    const auto doa = estimateAzimuthGcc(phase_safe_frame, profile_);
    azimuth = doa.azimuth_deg;
    doa_confidence = doa.confidence;
  }

  std::vector<float> spatial_audio = raw_mono;
  if (azimuth.has_value() && doa_confidence >= 0.1F) {
    spatial_audio = delayAndSum(phase_safe_frame, profile_, *azimuth);
  }
  const float self_echo = estimateSelfEcho(spatial_audio, playback_reference);
  const auto echo_cancelled = echo_canceller_.process(spatial_audio, playback_reference);
  std::vector<float> target_audio = noise_suppressor_.process(echo_cancelled, pre_activity.active);
  const VoiceActivity activity = post_vad_.process(target_audio);
  const float clarity = estimateClarity(activity.snr_db);

  FrontendOutput output;
  output.stamp_ms = frame.stamp_ms;
  output.target_audio = target_audio;
  if (!activity.active) {
    return output;
  }

  AcousticTrack track;
  track.track_id = tracker_.assign(frame.stamp_ms, azimuth);
  track.stamp_ms = frame.stamp_ms;
  track.voice_activity = activity.active;
  track.speech_probability = activity.probability;
  track.clarity = clarity;
  track.azimuth_deg = azimuth;
  track.overlap_probability = estimateOverlap(phase_safe_frame);
  track.self_echo_probability = self_echo;
  track.features = extractFeatures(target_audio, frame.sample_rate_hz);

  output.target_track_id = track.track_id;
  output.tracks.push_back(track);
  return output;
}

std::int64_t nowMs()
{
  const auto now = std::chrono::system_clock::now().time_since_epoch();
  return std::chrono::duration_cast<std::chrono::milliseconds>(now).count();
}

}  // namespace voice_detection

File: voice_detection/frontend.py
from __future__ import annotations

import time

import numpy as np

from .aec import LinearEchoCanceller
from .beamforming import delay_and_sum, mixdown
from .doa import estimate_azimuth_gcc
from .features import extract_features
from .noise import (
    AdaptiveNoiseSuppressor,
    estimate_clarity_from_snr,
    estimate_overlap_probability,
    estimate_self_echo_probability,
    remove_dc,
)
from .tracker import SourceTracker
from .types import AcousticTrack, AudioFrame, FrontendOutput, MicrophoneProfile
from .vad import EnergyVad


class CocktailFrontend:
    """Low-latency acoustic frontend for one dominant voice per audio block.

    The first implementation always provides VAD, clarity and target mixdown.
    With a calibrated array profile it adds GCC-PHAT DOA and delay-and-sum
    beamforming. Neural separation can be attached behind `FrontendOutput`
    without changing ROS4HRI or attention interfaces.
    """

    def __init__(self, profile: MicrophoneProfile, vad_calibration_frames: int = 0):
        self.profile = profile
        self.pre_vad = EnergyVad(calibration_frames=vad_calibration_frames)
        self.post_vad = EnergyVad(calibration_frames=vad_calibration_frames)
        self.noise_suppressor = AdaptiveNoiseSuppressor()
        self.echo_canceller = LinearEchoCanceller()
        self.tracker = SourceTracker()

    def process(self, frame: AudioFrame, playback_reference: np.ndarray | None = None) -> FrontendOutput:
        samples = remove_dc(frame.samples)
        # Keep the original, stable frontend as the default signal path.  In
        # particular, do not switch channels or alter gain independently for
        # every 20 ms block: both operations move the adaptive VAD noise floor
        # and change the waveform presented to ASR.
        raw_mono = mixdown(samples, self.profile.doa_channel_indices or None)
        pre_vad = self.pre_vad.process(raw_mono)

        azimuth_deg = None
        doa_confidence = 0.0
        if pre_vad.active and self.profile.supports_doa and self.profile.doa_channel_indices:
            doa = estimate_azimuth_gcc(
                samples=samples,
                sample_rate_hz=frame.sample_rate_hz,
                mic_positions_m=self.profile.mic_positions_m,
                channel_indices=self.profile.doa_channel_indices,
            )
            azimuth_deg = doa.azimuth_deg
            doa_confidence = doa.confidence

        if azimuth_deg is not None and doa_confidence >= 0.1:
            spatial_audio = delay_and_sum(
                samples=samples,
                sample_rate_hz=frame.sample_rate_hz,
                mic_positions_m=self.profile.mic_positions_m,
                channel_indices=self.profile.doa_channel_indices,
                azimuth_deg=azimuth_deg,
            )
        else:
            spatial_audio = raw_mono
        raw_echo = estimate_self_echo_probability(spatial_audio, playback_reference)
        echo_cancelled = self.echo_canceller.process(spatial_audio, playback_reference)
        self_echo = estimate_self_echo_probability(echo_cancelled, playback_reference)
        target_audio = self.noise_suppressor.process(echo_cancelled, pre_vad.active)
        vad = self.post_vad.process(target_audio)
        clarity = estimate_clarity_from_snr(vad.snr_db)

        track_id = self.tracker.assign(frame.stamp_ms, azimuth_deg) if vad.active else ""
        features = extract_features(target_audio, frame.sample_rate_hz)
        overlap = estimate_overlap_probability(samples) if vad.active else 0.0
        tracks = ()
        if vad.active:
            tracks = (
                AcousticTrack(
                    track_id=track_id,
                    stamp_ms=frame.stamp_ms,
                    voice_activity=vad.active,
                    speech_probability=vad.probability,
                    clarity=clarity,
                    azimuth_deg=azimuth_deg,
                    overlap_probability=overlap,
                    self_echo_probability=self_echo,
                    raw_echo_probability=raw_echo,
                    features=features,
                ),
            )
        return FrontendOutput(
            stamp_ms=frame.stamp_ms,
            tracks=tracks,
            target_audio=target_audio.astype(np.float32),
            target_track_id=track_id,
            vad_active=vad.active,
            vad_probability=vad.probability,
            rms_dbfs=vad.rms_dbfs,
            noise_floor_dbfs=vad.noise_floor_dbfs,
            snr_db=vad.snr_db,
            agc_gain_db=0.0,
        )


def now_ms() -> int:
    return int(time.time() * 1000)
